### DISCUSSION TITLE
Error running shebang with carriage return: is it xonsh or not?

### DISCUSSION THREAD

**Discussion by eugenesvk** [xonsh/xonsh#5615] [Q&A] [answered]
upd: found the issue, xonsh wasn't expanding relative paths, so that's why could only run absolute paths

Tried to run a Win-generated xonsh file on a Mac and got a `env: xonsh\r: no such file or directory` error since the first shebang line ended with a carriage return.

Is this a xonsh parsing bug or something more broader?
Suprisingly `./ttt.xsh` fails, but launching via the full `~/path/to/ttt.xsh`

(other shells fail with both variants)

Test
[ttt.txt](https://github.com/user-attachments/files/16331039/ttt.txt)
- download
- change extension to `.xsh`
- open xonsh shell
- `chmod +x ttt.xsh`
- try runing as `./ttt.xsh` - should fail
- try runing as `/Users/path/to/ttt.xsh` - should succeed

**Comment by anki-code** [comment]
First of all please form repeatable test-case e.g.
```xsh
cd /tmp
echo @("""#!/usr/bin/env xonsh
echo 1
""") > file.xsh
chmod +x file.xsh
./file.xsh
/tmp/file.xsh
```

**Reply by eugenesvk** [reply]
thanks, added

**Comment by bestlem** [comment]
It is probably a Windows/Unix issue

By default Windows ends lines in CRLF whilst Unix uses CR.

The best fix is use an editor in Windows that uses CR only
In Unix there is often a script dos2unix that will convert the line endings.

Or use tr
```
tr -d '\r' <filename >filename.unix
```

**Reply by eugenesvk** [reply]
having to manually adjust these things is unfortunately the worst fix. If it's xonsh that's parsing files when they're invoked within xonsh, then this can be fixed at the parsing stage and not require constant vigilance

**Reply by bestlem** [reply]
Unless xonsh is your default shell it is not xonsh reading the file it is the default shell doing the reading.

This is a problem common for all shells. Windows files are not fully readable on Unix. It has been an issue  since Unix was created, different OS use different line endings and you have to convert - or now the easiest way is use Windows editors that do not use Windows line endings - which is all developers editors I think - I would be interested if there are ones that don't do this.

**Reply by eugenesvk** [reply]
I've changed it to be the default shell, but it didn't help

Though I've discovered that the file is always parsed by xonsh even if it's not the default, just for some reason there is a different parsing applied to the `./` and the full form

**Reply by eugenesvk** [reply]
I've set my editor not to use those dumb CRs as well, but then sometimes they still slip through since some other apps continue to create CRs

**Comment by eugenesvk** [comment, marked as answer]
answer https://github.com/xonsh/xonsh/pull/5617
